rust/src/bin/multiplex.rs

```diff
@@ -0,0 +1,352 @@
+}
+
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct TimeSpec {
+    sec: usize,
+    nsec: usize,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct SockAddrIn {
+    pub sin_family: u16,
+    pub sin_port: u16,
+    pub sin_addr: u32,
+    pub sin_zero: [u8; 8],
+}
+
+#[repr(u16)]
+pub enum AddressFamily{
+    /// Unspecified
+    Unspecified = 0,
+    /// Unix domain sockets
+    Unix= 1,
+    /// Internet IP Protocol
+    Internet = 2,
+    /// Netlink
+     Netlink = 16,
+    /// Packet family
+     Packet = 17,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct SocketType;
+
+impl SocketType {
+    /// Stream
+    pub const Stream: u8 = 1;
+    /// Datagram
+    pub const Datagram: u8 = 2;
+    /// Raw
+    pub const Raw: u8 = 3;
+}
+
+
+fn create_socket(id :u16) -> (usize, SockAddrIn) {
+    let mut servaddr: SockAddrIn = SockAddrIn{
+        sin_family: 0,
+        sin_port: 0,
+        sin_addr: 0,
+        sin_zero: [0u8; 8],
+    };;
+    servaddr.sin_family = AddressFamily::Internet as u16;
+    servaddr.sin_port = id;
+    servaddr.sin_addr = 0;
+        
+    let listenfd = sys_socket(AddressFamily::Internet as usize, SocketType::Stream as usize, 0) as usize;
+ 
+    if listenfd as i32 == -1 {
+        println!("can create socket");
+    }
+ 
+    let reuseaddr = 1u32;
+    if sys_setsockopt(listenfd, 1, 2, &reuseaddr as *const u32 as usize, 4) == -1{
+        println!("setsockopt error");
+    }
+  
+    if sys_bind(listenfd, &servaddr as *const SockAddrIn as usize , 16) == -1{
+        println!("bind error");
+    }
+ 
+    if sys_listen(listenfd, 5) == -1{
+        println!("listen error");
+    }
+ 
+    return (listenfd, servaddr);
+}
+
+fn test_ppoll(){
+    let (listensocket, servaddr) = create_socket(18463);
+    println!("create socket {}  , listening on 0.0.0.0, port 8008, use telnet to connect", listensocket);
+
+    let std_fd = PollFd {
+        fd: 0,
+        events: PollEvents::IN | PollEvents::ERR,
+        revents: 0,
+    };
+
+    let socket_fd = PollFd {
+        fd: listensocket as u32,
+        events: PollEvents::IN | PollEvents::ERR,
+        revents: 0, 
+    };
+
+    let mut fds = [std_fd, socket_fd];
+    let timeout = TimeSpec {
+        sec: 0,
+        nsec: 5000,
+    };
+    const buf_sz: usize = 1024;
+    let mut buf_read = [0u8; buf_sz];
+	
+
+	let ret = sys_ppoll(fds.as_mut_ptr() as *const u32 as usize, 2, &timeout as *const TimeSpec as usize);
+	
+    if ret < 0{
+        println!("error happened");
+    } if ret == 0{
+    	println!("poll run out of time: {}ms ", timeout.nsec);
+    }else {
+        if (fds[0].revents & PollEvents::IN) != 0 {
+            println!("stdin is readable: ");
+            sys_read(0, buf_read.as_mut_ptr(), buf_sz);
+            println!("input: {}", String::from_utf8_lossy(&buf_read));
+        }
+
+        if (fds[1].revents & PollEvents::IN) != 0 {
+            println!("network is readable");
+            let client: SockAddrIn = SockAddrIn{
```

Comment:
WARN: This should be mutable!
(That's why I suggest syscall fn declarations should use strong type)